File: js/src/formulas/QRDecomposition.ts
import matrixScalaMultiplation from './matrixScalaMultiplation'
import MatrixSub from './MatrixSub'
import transpose from './transpose'

export default function QRDecomposition (mat: number[][]) {
  function sqrtDotProduct (row1: number[], row2: number[]) {
    let res = 0
    for (let i = 0; i < row1.length; i++) {
      res += row1[i] * row2[i]
    }

    return res ** 0.5
  }

  function dotProduct (row1: number[], row2: number[]) {
    let res = 0
    for (let i = 0; i < row1.length; i++) {
      res += row1[i] * row2[i]
    }

    return res
  }

  function normalizeMemoi (row: number[][], K: number) {
    return matrixScalaMultiplation(row, 1 / K)
  }

  const X = transpose(mat)

  const r = [] as number[][]
  for (let i = 0; i < X.length; i++) {
    r.push([])
    for (let j = 0; j < X[0].length; j++) {
      r[i].push(0)
    }
  }
  const Q = [] as number[][]
  for (let i = 0; i < X.length; i++) {
    Q.push([0])
  }

  let k = 0

  while (1) {
    r[k][k] = sqrtDotProduct(X[k], X[k])

    Q[k] = normalizeMemoi([X[k]], r[k][k])[0]

    for (let j = k + 1; j < X.length; j++) {
      r[k][j] = dotProduct(Q[k], X[j])
    }

    for (let j = k + 1; j < X.length; j++) {
      X[j] = MatrixSub([X[j]], matrixScalaMultiplation([Q[k]], r[k][j]))[0]
    }

    k += 1

    if (k === X.length) {
      return {
        Q: transpose(Q),
        R: r
      }
    }
  }
}

File: py/src/formulas/QRDecomposition.py
from .matrixScalaMultiplation import matrix_scala_multiplation
from .MatrixSub import matrix_sub
from .transpose import transpose

def QRDecomposition(mat: list[list[float]]) -> dict:
    def sqrt_dot_product(row1: list[float], row2: list[float]) -> float:
        res = 0
        for i in range(len(row1)):
            res += row1[i] * row2[i]
        return res ** 0.5

    def dot_product(row1: list[float], row2: list[float]) -> float:
        res = 0
        for i in range(len(row1)):
            res += row1[i] * row2[i]
        return res

    def normalize_memoi(row: list[list[float]], K: float) -> list[list[float]]:
        return matrix_scala_multiplation(row, 1 / K)

    X = transpose(mat)

    r = []
    for i in range(len(X)):
        r.append([0] * len(X[0]))
    Q = []
    for i in range(len(X)):
        Q.append([0] * len(X[0]))

    k = 0

    while True:
        r[k][k] = sqrt_dot_product(X[k], X[k])
        Q[k] = normalize_memoi([X[k]], r[k][k])[0]

        for j in range(k + 1, len(X)):
            r[k][j] = dot_product(Q[k], X[j])

        for j in range(k + 1, len(X)):
            X[j] = matrix_sub([X[j]], matrix_scala_multiplation([Q[k]], r[k][j]))[0]

        k += 1

        if k == len(X):
            return {
                'Q': transpose(Q),
                'R': r
            }